iml-services/iml-snapshot/src/client_monitor.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{Error, MonitorState};
+use iml_command_utils::wait_for_cmds_success;
+use iml_graphql_queries::snapshot as snapshot_queries;
+use iml_manager_client::{get_influx, graphql, Client};
+use iml_postgres::{sqlx, PgPool};
+use iml_tracing::tracing;
+use std::{collections::HashMap, fmt::Debug};
+use tokio::time::{Duration, Instant};
+
+pub async fn tick(
+    snapshot_client_counts: &mut HashMap<i32, MonitorState>,
+    pool: PgPool,
+) -> Result<(), Error> {
+    let client: Client = iml_manager_client::get_client()?;
+    let client_2 = client.clone();
+
+    let query = iml_influx::filesystems::query();
+    let stats_fut =
+        get_influx::<iml_influx::filesystems::InfluxResponse>(client, "iml_stats", query.as_str());
+
+    let influx_resp = stats_fut.await?;
+    let stats = iml_influx::filesystems::Response::from(influx_resp);
+
+    tracing::debug!("Influx stats: {:?}", stats);
+
+    let snapshots =
+        sqlx::query!("SELECT id, filesystem_name, snapshot_name, snapshot_fsname FROM snapshot")
+            .fetch_all(&pool)
+            .await?;
+    tracing::debug!("Fetched {} snapshots from DB", snapshots.len());
+
+    for snapshot in snapshots {
+        let snapshot_id = snapshot.id;
+        let snapshot_stats = match stats.get(&snapshot.snapshot_fsname) {
+            Some(x) => x,
+            None => continue,
+        };
+
+        let monitor_state = snapshot_client_counts.get_mut(&snapshot_id);
+        let clients = snapshot_stats.clients.unwrap_or(0);
+
+        match monitor_state {
+            Some(MonitorState::Monitoring(prev_clients)) => {
+                tracing::debug!(
+                    "Monitoring. Snapshot {}: {} clients (previously {} clients)",
+                    &snapshot.snapshot_fsname,
+                    clients,
+                    prev_clients,
+                );
+                if *prev_clients > 0 && clients == 0 {
+                    tracing::trace!("counting down for job");
+                    let instant = Instant::now() + Duration::from_secs(5 * 60);
+                    monitor_state.map(|s| *s = MonitorState::CountingDown(instant));
+                } else {
+                    *prev_clients = clients;
+                }
+            }
+            Some(MonitorState::CountingDown(when)) => {
+                tracing::debug!(
+                    "Counting down. Snapshot {}: 0 clients (previously {} clients)",
+                    &snapshot.snapshot_fsname,
+                    clients
+                );
+                if clients > 0 {
+                    tracing::trace!("changing MonitorState");
+                    monitor_state.map(|s| *s = MonitorState::Monitoring(clients));
+                } else if Instant::now() >= *when {
+                    tracing::trace!("running the job");
+                    monitor_state.map(|s| *s = MonitorState::Monitoring(0));
+
+                    let query = snapshot_queries::unmount::build(
```

Comment:
Are we sure we want to 1) unmount snapshots before destroying (and not just wait) 2) duplicate the unmount feature here?